test/v200/utils/common/project_test_utils.go

```diff
@@ -0,0 +1,244 @@
+}
+
+// AddStarterProject adds a starter project of the specified type, with random attributes, to the devfile schema
+func (testDevfile *TestDevfile) AddStarterProject(projectType schema.ProjectSourceType) string {
+	starterProject := testDevfile.createStarterProject(projectType)
+	testDevfile.SetStarterProjectValues(&starterProject)
+	return starterProject.Name
+}
+
+// createProject creates a project of a specified type with only required attributes set
+func (testDevfile *TestDevfile) createProject(projectType schema.ProjectSourceType) schema.Project {
+	project := schema.Project{}
+	project.Name = GetRandomUniqueString(GetRandomNumber(8, 63), true)
+	LogInfoMessage(fmt.Sprintf("Create Project Name: %s", project.Name))
+
+	if projectType == schema.GitProjectSourceType {
+		project.Git = createGitProject()
+	} else if projectType == schema.GitHubProjectSourceType {
+		project.Github = createGithubProject()
+	} else if projectType == schema.ZipProjectSourceType {
+		project.Zip = createZipProject()
+	}
+	testDevfile.projectAdded(project)
+	return project
+}
+
+// generateStarterProject creates a starter project of a specified type with only required attributes set
+func (testDevfile *TestDevfile) createStarterProject(projectType schema.ProjectSourceType) schema.StarterProject {
+	starterProject := schema.StarterProject{}
+	starterProject.Name = GetRandomUniqueString(GetRandomNumber(8, 63), true)
+	LogInfoMessage(fmt.Sprintf("Create StarterProject Name: %s", starterProject.Name))
+
+	if projectType == schema.GitProjectSourceType {
+		starterProject.Git = createGitProject()
+	} else if projectType == schema.GitHubProjectSourceType {
+		starterProject.Github = createGithubProject()
+	} else if projectType == schema.ZipProjectSourceType {
+		starterProject.Zip = createZipProject()
+	}
+	testDevfile.starterProjectAdded(starterProject)
+	return starterProject
+
+}
+
+// createGitProject creates a git project structure with mandatory attributes set
+func createGitProject() *schema.GitProjectSource {
+	project := schema.GitProjectSource{}
+	project.Remotes = getRemotes()
+	return &project
+}
+
+// createGithubProject creates a github project structure with mandatory attributes set
+func createGithubProject() *schema.GithubProjectSource {
+	project := schema.GithubProjectSource{}
+	project.Remotes = getRemotes()
+	return &project
+}
+
+// createZipProject creates a zip project structure
+func createZipProject() *schema.ZipProjectSource {
+	project := schema.ZipProjectSource{}
+	return &project
+}
+
+// SetProjectValues sets project attributes, common to all projects, to random values.
+func (testDevfile *TestDevfile) SetProjectValues(project *schema.Project) {
+
+	if GetBinaryDecision() {
+		project.ClonePath = "./" + GetRandomString(GetRandomNumber(4, 12), false)
+		LogInfoMessage(fmt.Sprintf("Set ClonePath : %s", project.ClonePath))
+	}
+
+	if GetBinaryDecision() {
+		var sparseCheckoutDirs []string
+		numDirs := GetRandomNumber(1, 6)
```

Comment:
Same as above, do we need to use a random number of directories here, or can we use a fixed number (and/or have separate test cases if we really need to test different numbers of directories), and keep things reproducible between test runs?

Reply:
Same response as above. Also I should say the tests are set up to run the same test 5 times to get a nice variety of random values. Maybe we can discuss this at the next team meeting?